assets/model_monitoring/components/src/generation_safety_quality/input_schema_adaptor/run.py

```diff
@@ -5,10 +5,55 @@
 
 import argparse
 
+from pyspark.sql import DataFrame
+from pyspark.sql.functions import from_json
 from shared_utilities.io_utils import (
+    init_spark,
     save_spark_df_as_mltable,
     try_read_mltable_in_spark_with_error,
 )
+from shared_utilities.momo_exceptions import InvalidInputError
+from model_data_collector_preprocessor.spark_run import _convert_complex_columns_to_json_string
+from shared_utilities.constants import GENAI_ROOT_SPAN_SCHEMA_COLUMN, GENAI_TRACE_ID_SCHEMA_COLUMN
+
+
+def _adapt_input_data_schema(df: DataFrame) -> DataFrame:
+    """Adapt the input dataframe schema to fit GSQ input schema."""
+    df_field_names = df.schema.fieldNames()
+
+    # check if we need to adapt the schema
+    if GENAI_ROOT_SPAN_SCHEMA_COLUMN not in df_field_names and GENAI_TRACE_ID_SCHEMA_COLUMN not in df_field_names:
+        return df
+
+    spark = init_spark()
+    try:
+        input_schema = spark.read.json(df.rdd.map(lambda row: row.input), mode="FAILFAST").schema
+        output_schema = spark.read.json(df.rdd.map(lambda row: row.output), mode="FAILFAST").schema
+    except Exception as ex:
+        if "Malformed records are detected in schema inference. Parse Mode: FAILFAST." in str(ex):
+            raise InvalidInputError(
+                "Failed to parse the input/output column json string for the trace logs provided."
+                " The input/output columns in the inputted production data are not in a parseable" +
+                " json string format. Please double-check the data columns are being passed or" +
+                " stored correctly."
+            )
+        raise ex
+
+    df = df.withColumns({
+        'temp_input': from_json(df.input, input_schema),
+        'temp_output': from_json(df.output, output_schema),
+    }).select('trace_id', 'temp_input.*', 'temp_output.*', 'root_span')
+
+    # flatten unpacked json columns to json_string if necessary
+    df = _convert_complex_columns_to_json_string(df)
+
+    df.show(truncate=False)
+    if df.isEmpty():
```

Comment:
I suppose we have a 1:1 row mapping between the original production_data_df and the adapted df? If so, should we do the df empty check against production_data_df instead of the adapted df?

Reply:
We already have those validations when reading from mltable. I actually will remove this isEmpty() check since I added it before including the 'trace_id' and 'root_span' columns in the output df. I was trying to check if the input/output columns have no nested data but that will just be passed down stream and fail in the gsq.